# Final Project

Group 4: Jiaqi Bai, Somender Chaudhary, Shreyas Deo, Andrew Levy, Wenqwing Qin, Matias Roca-Guiulfo, and Alexis Yang

## Ask 1 - Data Collection (10%)

Use the JHU datasets website, collect the following dataset 
- Cases and Deaths (04/12/2020 – 11/15/2021)
- Vaccine and People Vaccinated (12/10/2020 – 11/15/2021)
- Testing

The readme/dictionary files are available for all datasets Cases and Deaths, Vaccine and Testing.

Acquire daily U.S. report. Perform initial exploration to make sure it is suitable for dimensional modeling and analytical analysis.

Expectation: Provide a list of your data sources, The links to the data sources and links to any documentations you used.

Links
- JHU Datasets: https://coronavirus.jhu.edu/about/how-to-use-our-data
- Cases: https://github.com/CSSEGISandData/COVID-19/tree/master/csse_covid_19_data/csse_covid_19_daily_reports_us
- Deaths: https://github.com/govex/COVID-19/blob/master/data_tables/vaccine_data/us_data/time_series/vaccine_data_us_timeline.csv & https://raw.githubusercontent.com/govex/COVID-
- Vaccine: https://github.com/govex/COVID-19/blob/master/data_tables/vaccine_data/us_data/time_series/vaccine_data_us_timeline.csv
- People Vaccinated: https://raw.githubusercontent.com/govex/COVID-19/master/data_tables/vaccine_data/us_data/time_series/people_vaccinated_us_timeline.csv
- Testing: https://raw.githubusercontent.com/govex/COVID-19/master/data_tables/testing_data/time_series_covid19_US.csv
Readme/dictionary files
- https://github.com/CSSEGISandData/COVID-19/blob/master/csse_covid_19_data/README.md
- https://github.com/govex/COVID-19/blob/master/data_tables/vaccine_data/us_data/data_dictionary.csv
- https://github.com/govex/COVID-19/blob/master/data_tables/testing_data/data_dictionary.csv

#### Collecting Cases and Deaths Dataset 

In [6]:
!wc -l cases_deaths.csv # printing the number of lines presents in the cases_deaths.csv file

34471 cases_deaths.csv


In [7]:
!head cases_deaths.csv # printing the first 10 lines of the cases_deaths.csv file

Province_State,Country_Region,Last_Update,Lat,Long_,Confirmed,Deaths,Recovered,Active,FIPS,Incident_Rate,Total_Test_Results,People_Hospitalized,Case_Fatality_Ratio,UID,ISO3,Testing_Rate,Hospitalization_Rate
Alabama,US,2021-01-02 05:30:44,32.3182,-86.9023,365747,4872,202137.0,158738.0,1.0,7459.375895463867,3275341.0,,1.3320683423240656,84000001.0,USA,66800.27369964625,
Alaska,US,2021-01-02 05:30:44,61.3707,-152.4044,47019,206,7165.0,39615.0,2.0,6422.844801071706,1275750.0,,0.4384284680543141,84000002.0,USA,174391.1857780451,
American Samoa,US,2021-01-02 05:30:44,-14.271,-170.132,0,0,,0.0,60.0,0.0,2140.0,,,16.0,ASM,3846.084721698028,
Arizona,US,2021-01-02 05:30:44,33.7298,-111.4312,530267,9015,76934.0,444318.0,4.0,7285.171274003372,5155330.0,,1.7000869373353424,84000004.0,USA,39551.860582022906,
Arkansas,US,2021-01-02 05:30:44,34.9697,-92.3731,229442,3711,199247.0,26484.0,5.0,7602.945718144717,2051488.0,,1.6174022192972517,84000005.0,USA,67979.49767446793,
California,US,2021-01-02 05:30:

In [8]:
!head -5 cases_deaths.csv | csvlook # rendering the cases_deaths.csv file into a readable format 

| Province_State | Country_Region |         Last_Update |      Lat |     Long_ | Confirmed | Deaths | Recovered |  Active | FIPS | Incident_Rate | Total_Test_Results | People_Hospitalized | Case_Fatality_Ratio |        UID | ISO3 | Testing_Rate | Hospitalization_Rate |
| -------------- | -------------- | ------------------- | -------- | --------- | --------- | ------ | --------- | ------- | ---- | ------------- | ------------------ | ------------------- | ------------------- | ---------- | ---- | ------------ | -------------------- |
| Alabama        | US             | 2021-01-02 05:30:44 |  32.318… |  -86.902… |   365,747 |  4,872 |   202,137 | 158,738 |    1 |    7,459.376… |          3,275,341 |                     |              1.332… | 84,000,001 | USA  |  66,800.274… |                      |
| Alaska         | US             | 2021-01-02 05:30:44 |  61.371… | -152.404… |    47,019 |    206 |     7,165 |  39,615 |    2 |    6,422.845… |          1,275,750 |                     | 

In [9]:
!head -n 1000 cases_deaths.csv | csvstat # printing the descriptive statistics for the first 1000 lines of the cases_deaths.csv file 

  1. "Province_State"

	Type of data:          Text
	Contains null values:  False
	Unique values:         58
	Longest value:         24 characters
	Most common values:    Alabama (18x)
	                       Alaska (18x)
	                       American Samoa (18x)
	                       Arizona (18x)
	                       Arkansas (18x)

  2. "Country_Region"

	Type of data:          Text
	Contains null values:  False
	Unique values:         1
	Longest value:         2 characters
	Most common values:    US (999x)

  3. "Last_Update"

	Type of data:          DateTime
	Contains null values:  False
	Unique values:         29
	Smallest value:        2021-01-02 05:30:44
	Largest value:         2021-01-19 09:30:34
	Most common values:    2021-01-02 05:30:44 (58x)
	                       2021-01-03 05:30:33 (58x)
	                       2021-01-04 05:30:45 (58x)
	                       2021-01-05 05:30:33 (58x)
	                       2021-01-06 05:30:39 (58x)

  4. "Lat"

	Type of data:

#### Collecting Vaccine Dataset

In [10]:
!wget https://raw.githubusercontent.com/govex/COVID-19/master/data_tables/vaccine_data/us_data/time_series/vaccine_data_us_timeline.csv

--2021-11-27 17:37:04--  https://raw.githubusercontent.com/govex/COVID-19/master/data_tables/vaccine_data/us_data/time_series/vaccine_data_us_timeline.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.111.133, 185.199.108.133, 185.199.109.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.111.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 10348810 (9.9M) [text/plain]
Saving to: ‘vaccine_data_us_timeline.csv’

vaccine_data_us_tim 100%[===================>]   9.87M  --.-KB/s    in 0.09s   

2021-11-27 17:37:04 (109 MB/s) - ‘vaccine_data_us_timeline.csv’ saved [10348810/10348810]



In [11]:
!mv vaccine_data_us_timeline.csv vaccine.csv # changing the name of the orginal downloaded csv file into vaccine.csv

In [12]:
!wc -l vaccine.csv # printing the number of lines presents in the vaccine.csv file

116249 vaccine.csv


In [13]:
!head vaccine.csv | csvcut -n # printing the first 10 lines of the vaccine.csv file

  1: Province_State
  2: Date
  3: Vaccine_Type
  4: FIPS
  5: Country_Region
  6: Lat
  7: Long_
  8: Doses_alloc
  9: Doses_shipped
 10: Doses_admin
 11: Stage_One_Doses
 12: Stage_Two_Doses
 13: Combined_Key


In [14]:
!head -5 vaccine.csv | csvlook # rendering the vaccine.csv file into a readable format 

| Province_State |       Date | Vaccine_Type | FIPS | Country_Region |     Lat |    Long_ | Doses_alloc | Doses_shipped | Doses_admin | Stage_One_Doses | Stage_Two_Doses | Combined_Key |
| -------------- | ---------- | ------------ | ---- | -------------- | ------- | -------- | ----------- | ------------- | ----------- | --------------- | --------------- | ------------ |
| Alabama        | 2020-12-10 | All          | True | US             | 32.318… | -86.902… |             |               |             |                 |                 | Alabama, US  |
| Alabama        | 2020-12-10 | Moderna      | True | US             | 32.318… | -86.902… |             |               |             |                 |                 | Alabama, US  |
| Alabama        | 2020-12-10 | Pfizer       | True | US             | 32.318… | -86.902… |             |               |             |                 |                 | Alabama, US  |
| Alabama        | 2020-12-11 | All          | True | US         

In [15]:
!head -n 1000 vaccine.csv | csvstat # printing the descriptive statistics for the first 1000 lines of the vaccine.csv file 

  1. "Province_State"

	Type of data:          Text
	Contains null values:  False
	Unique values:         1
	Longest value:         7 characters
	Most common values:    Alabama (999x)

  2. "Date"

	Type of data:          Date
	Contains null values:  False
	Unique values:         198
	Smallest value:        2020-12-10
	Largest value:         2021-06-25
	Most common values:    2021-03-15 (6x)
	                       2021-03-16 (6x)
	                       2021-03-17 (6x)
	                       2021-03-18 (6x)
	                       2021-03-19 (6x)

  3. "Vaccine_Type"

	Type of data:          Text
	Contains null values:  False
	Unique values:         6
	Longest value:         10 characters
	Most common values:    All (198x)
	                       Moderna (198x)
	                       Pfizer (198x)
	                       Unassigned (194x)
	                       Janssen (109x)

  4. "FIPS"

	Type of data:          Boolean
	Contains null values:  False
	Unique values:         1
	Most

#### Collecting People Vaccinated Dataset

In [16]:
!wget https://raw.githubusercontent.com/govex/COVID-19/master/data_tables/vaccine_data/us_data/time_series/people_vaccinated_us_timeline.csv

--2021-11-27 17:37:20--  https://raw.githubusercontent.com/govex/COVID-19/master/data_tables/vaccine_data/us_data/time_series/people_vaccinated_us_timeline.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.110.133, 185.199.109.133, 185.199.108.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.110.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1585905 (1.5M) [text/plain]
Saving to: ‘people_vaccinated_us_timeline.csv’

people_vaccinated_u 100%[===================>]   1.51M  --.-KB/s    in 0.01s   

2021-11-27 17:37:20 (105 MB/s) - ‘people_vaccinated_us_timeline.csv’ saved [1585905/1585905]



In [17]:
!mv people_vaccinated_us_timeline.csv people_vaccinated.csv # changing the name of the orginal downloaded csv file into people_vaccinated.csv

In [18]:
!wc -l people_vaccinated.csv # printing the number of lines presents in the people_vaccinated.csv file

21433 people_vaccinated.csv


In [19]:
!head people_vaccinated.csv | csvcut -n # printing the first 10 lines of the people_vaccinated.csv file

  1: FIPS
  2: Province_State
  3: Country_Region
  4: Date
  5: Lat
  6: Long_
  7: Combined_Key
  8: People_Fully_Vaccinated
  9: People_Partially_Vaccinated


In [20]:
!head -5 people_vaccinated.csv | csvlook # rendering the people_vaccinated.csv file into a readable format 

| FIPS | Province_State | Country_Region |       Date |     Lat |    Long_ | Combined_Key | People_Fully_Vaccinated | People_Partially_Vaccinated |
| ---- | -------------- | -------------- | ---------- | ------- | -------- | ------------ | ----------------------- | --------------------------- |
| True | Alabama        | US             | 2020-12-10 | 32.318… | -86.902… | Alabama, US  |                         |                             |
| True | Alabama        | US             | 2020-12-11 | 32.318… | -86.902… | Alabama, US  |                         |                             |
| True | Alabama        | US             | 2020-12-12 | 32.318… | -86.902… | Alabama, US  |                         |                             |
| True | Alabama        | US             | 2020-12-13 | 32.318… | -86.902… | Alabama, US  |                         |                             |


In [21]:
!head -n 1000 vaccine.csv | csvstat # printing the descriptive statistics for the first 1000 lines of the people_vaccinated.csv file 

  1. "Province_State"

	Type of data:          Text
	Contains null values:  False
	Unique values:         1
	Longest value:         7 characters
	Most common values:    Alabama (999x)

  2. "Date"

	Type of data:          Date
	Contains null values:  False
	Unique values:         198
	Smallest value:        2020-12-10
	Largest value:         2021-06-25
	Most common values:    2021-03-15 (6x)
	                       2021-03-16 (6x)
	                       2021-03-17 (6x)
	                       2021-03-18 (6x)
	                       2021-03-19 (6x)

  3. "Vaccine_Type"

	Type of data:          Text
	Contains null values:  False
	Unique values:         6
	Longest value:         10 characters
	Most common values:    All (198x)
	                       Moderna (198x)
	                       Pfizer (198x)
	                       Unassigned (194x)
	                       Janssen (109x)

  4. "FIPS"

	Type of data:          Boolean
	Contains null values:  False
	Unique values:         1
	Most

#### Collecting Testing Dataset 

In [22]:
!wget https://raw.githubusercontent.com/govex/COVID-19/master/data_tables/testing_data/time_series_covid19_US.csv

--2021-11-27 17:37:37--  https://raw.githubusercontent.com/govex/COVID-19/master/data_tables/testing_data/time_series_covid19_US.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.110.133, 185.199.109.133, 185.199.108.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.110.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 2271759 (2.2M) [text/plain]
Saving to: ‘time_series_covid19_US.csv’

time_series_covid19 100%[===================>]   2.17M  --.-KB/s    in 0.02s   

2021-11-27 17:37:37 (89.5 MB/s) - ‘time_series_covid19_US.csv’ saved [2271759/2271759]



In [23]:
!mv time_series_covid19_US.csv testing.csv # changing the name of the orginal downloaded csv file into testing.csv

In [24]:
!wc -l testing.csv # printing the number of lines presents in the testing.csv file

35565 testing.csv


In [25]:
!head testing.csv | csvcut -n # printing the first 10 lines of the testing.csv file

  1: date
  2: state
  3: cases_conf_probable
  4: cases_confirmed
  5: cases_probable
  6: tests_viral_positive
  7: tests_viral_negative
  8: tests_viral_total
  9: tests_antigen_positive
 10: tests_antigen_total
 11: people_viral_positive
 12: people_viral_total
 13: people_antigen_positive
 14: people_antigen_total
 15: encounters_viral_total
 16: tests_combined_total


In [26]:
!head -5 testing.csv | csvlook # rendering the testing.csv file into a readable format 

|       date | state | cases_conf_probable | cases_confirmed | cases_probable | tests_viral_positive | tests_viral_negative | tests_viral_total | tests_antigen_positive | tests_antigen_total | people_viral_positive | people_viral_total | people_antigen_positive | people_antigen_total | encounters_viral_total | tests_combined_total |
| ---------- | ----- | ------------------- | --------------- | -------------- | -------------------- | -------------------- | ----------------- | ---------------------- | ------------------- | --------------------- | ------------------ | ----------------------- | -------------------- | ---------------------- | -------------------- |
| 2020-03-06 | AK    |                     |                 |                |                      |                      |                 8 |                        |                     |                       |                    |                         |                      |                        |                   

In [27]:
!head -n 1000 testing.csv | csvstat # printing the descriptive statistics for the first 1000 lines of the testing.csv file 

  1. "date"

	Type of data:          Date
	Contains null values:  False
	Unique values:         600
	Smallest value:        2020-03-06
	Largest value:         2021-10-26
	Most common values:    2020-03-07 (2x)
	                       2020-03-08 (2x)
	                       2020-03-09 (2x)
	                       2020-03-10 (2x)
	                       2020-03-11 (2x)

  2. "state"

	Type of data:          Text
	Contains null values:  False
	Unique values:         2
	Longest value:         2 characters
	Most common values:    AK (600x)
	                       AL (399x)

  3. "cases_conf_probable"

	Type of data:          Number
	Contains null values:  True (excluded from calculations)
	Unique values:         872
	Smallest value:        0
	Largest value:         518234
	Sum:                   103854382
	Mean:                  105115.771255
	Median:                60399
	StDev:                 140925.470189
	Most common values:    None (11x)
	                       0 (6x)
	               

## Ask 2 - Surveying and Wrangling Data (15%)

Acquire the data, survey it and wrangle it to suitable format. You may use one or more of the following methods for acquiring and wrangling the data: 
- Unix command line tools (e.g. csvkit)
- Relational databases 
- Spark
- You may also use wrangling tool (e.g. Trifacta)

Describe any concerns with the data and changes you expect to overcome. 

Expectations: document your findings in the code. Make sure you describe each step in a complete and concise manner. 

## Ask 3 - Attestation (-10% for non-compliance)

All project team members must contribute meaningfully to the project and have complete knowledge on any part of it.

Each group member MUST attest to his/her individual contributions in coding, and presentation write up. 

Penalty: -10% for non-compliance

## Ask 4 - Analytical Questions (15%)

Communicate among your group members and produce analytical questions you want to answer with the data. Minimum of 4 major questions are required. 

You may consider me as your Sponsor. Arrange a 30 min meeting with me to discuss this part, if necessary. 

Expectation: A list of analytical questions that your model can answer. Check https://coronavirus.jhu.edu
- ”tracking” -> “Critical Trends” or U.S. Map 
- “Vaccines” -> “Tracker” for ideas 

## Ask 5 - Designing a Dimensional Database (25%)

Design a dimensional database to cover “Cases and Deaths”, “Vaccinations” and “Testing” datasets (Daily and for U.S.). I expect that your design will allow all other groups to answer most of their questions. 

Be precise and document your design correctly. 

Expectation: A star schema diagram detailing the design. Would be nice to produce an image of the design and added it to your code.

## Ask 6 - Writing Queries for Analytical Questions (35%) 

Write queries to answer your analytical questions.

Produce charts and make sure to include your code or a copy of chart (if you use outside Jupyter notebook tools). 

You are free to use any tool for visualization (e.g. tableau, ggplot, matplotlib,..).

Expectation: set of queries, results and charts. Those should be compatible with your presentation.